torch/csrc/distributed/rpc/rref_context.h

```diff
@@ -47,26 +47,21 @@ class RRefContext {
   }
 
   // create a ``UserRRef`` owned by the worker ``ownerId``
-  template <typename T>
-  std::shared_ptr<UserRRef<T>> createUserRRef(worker_id_t ownerId);
+  std::shared_ptr<UserRRef> createUserRRef(worker_id_t ownerId, const TypePtr& type);
 
   // Convert an RRefForkData into an RRef. This RRef could be user or owner.
   // This RRef could have already existed before, or could be created in this
   // method.
-  template <typename T>
-  std::shared_ptr<RRef> getOrCreateRRef(const RRefForkData& rfd);
+  std::shared_ptr<RRef> getOrCreateRRef(const RRefForkData& rfd, c10::optional<TypePtr> type={});
 
   // Get the ``OwnerRRef`` of id ``rrefId``. If it does not exist, create a new
   // one.
-  template <typename T>
-  std::shared_ptr<OwnerRRef<T>> getOrCreateOwnerRRef(const RRefId& rrefId);
+  std::shared_ptr<OwnerRRef> getOrCreateOwnerRRef(const RRefId& rrefId, c10::optional<TypePtr> opt_type={});
```

Comment:
What are the cases where we don't need to pass in the type?

Reply:
Basically when we need to purely just get the RRef instead of potentially creating it, we don't need to pass in the type, otherwise we will need to pass the type in.